py4j-python/src/py4j/clientserver.py

```diff
@@ -521,18 +522,27 @@ def wait_for_commands(self):
         self.close(reset)
 
     def _call_proxy(self, obj_id, input):
-        return_message = proto.ERROR_RETURN_MESSAGE
-        if obj_id in self.pool:
-            try:
-                method = smart_decode(input.readline())[:-1]
-                params = self._get_params(input)
-                return_value = getattr(self.pool[obj_id], method)(*params)
-                return_message = proto.RETURN_MESSAGE + proto.SUCCESS +\
-                    get_command_part(return_value, self.pool)
-            except Exception:
-                logger.exception("There was an exception while executing the "
-                                 "Python Proxy on the Python Side.")
-        return return_message
+        if obj_id not in self.pool:
+            return proto.RETURN_MESSAGE + proto.ERROR +\
+                get_command_part('Object ID unknown', self.pool)
+
+        try:
+            method = smart_decode(input.readline())[:-1]
+            params = self._get_params(input)
+            return_value = getattr(self.pool[obj_id], method)(*params)
+            return proto.RETURN_MESSAGE + proto.SUCCESS +\
+                get_command_part(return_value, self.pool)
+        except Exception as e:
+            logger.exception("There was an exception while executing the "
+                             "Python Proxy on the Python Side.")
+
+            if isinstance(e, proto.Py4JJavaError):
+                java_exception = e.java_exception
```

Comment:
Just to make sure I understand, wouldn't this hide the python proxy?

example:

Java.m1() calls Python.p1()
Python.p1() calls Java.m2()
Java.m2() raises FooBarException
Python.p1() raises a Py4JJavaError(value=FooBarException)
Java.m1() receives the FooBarException instance

m1() has almost no clue that it called p1() (there is probably a mention to a proxy in m2() stack trace).

Reply:
Yeah it does hide p1 from the stack trace, which is an unfortunate side-effect.

What I'm trying to support here is allowing Python to implement Java interfaces more usefully. e.g. let's say there is a interface:

  interface Stream {
    Object getNext() throws NoSuchElementException;
  }

Where NoSuchElementException is being used to signal the end of the stream. Right now there is no way to signal the stream end if you implement this interface in Py4J because all we can throw is a Py4JException.